keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -243,147 +233,162 @@ A `Bucket` is not deleted if it is bound to a `BucketRequest`.
+9. `accountID`: username/access-key for the object storage provider to uniquely identify the user with access to this bucket. This value is minted by the provisioner when the `BucketAccessRequest` omits a ServiceAccount. There are several use cases involving `accountID`: 
    - Not relevant if serviceAccount is supplied.
-   - For brownfield cases when there is already a storage-side principal that you want to use for access, the admin can supply this, the driver will (possibly) enable access for that principal to the bucket in question and will return credentials.
-   - For greenfield cases, this could be populated by the driver. Maybe so that the driver can request deletion of this principal during deprovisioning
-1. `parameters`:  A map of string:string key values.  Allows admins to control user and access provisioning by setting provisioner key-values. Copied from `BucketAccessClass`. 
-1. `accessGranted`: if true access has been granted to the bucket. If false then access to the bucket has not been granted.
+   - For brownfield cases when there is already a storage-side account id that you want to use for access, the admin can supply this, the driver will (possibly) enable access for that account id to the bucket in question and will return credentials.
+   - For greenfield cases, this could be populated by the driver so that it can request deletion of this account id during deprovisioning.
+10. `parameters`:  A map of string:string key values.  Allows admins to control user and access provisioning by setting provisioner key-values. Copied from `BucketAccessClass`. 
+11. `accessGranted`: if true access has been granted to the bucket. If false then access to the bucket has not been granted.
 
 #### BucketAccessClass
 
-An immutable, cluster-scoped, custom resource providing a way for admins to specify policies that may be used to access buckets. A `BucketAccessClass` (BAC) is required for both greenfield and brownfield use cases. Besides naming the provisioner, a BAC references a config map which is expected to define the access policy for a given provider. It is typical for these config maps to reside in each provisioner's namespace, though this is not required.  Unlike BucketClasses, there is no default BAC.
+An immutable, cluster-scoped, brownfield-only, custom resource providing a way for admins to specify policies used to access buckets. A `BucketAccessClass` (BAC) is referenced by a user [BucketAccessRequest](#bucketaccessrequest) and is used to populate the [BucketAccess](#bucketaccess) instance. A BAC references a config map which is expected to define the access policy for a given provider. It is typical for these config maps to reside in each provisioner's namespace, though this is not required. Unlike a [BucketClass](#bucketclass), there is no default BAC.
```

Comment:
Is this access policy about read and/or write access to certain prefixes within the bucket? Then I have the following comment:

This is a cluster-scoped resource, but why would an admin prescribe some access policies to choose from? After all, a Bucket belongs to the user that created the BucketRequest (or, for brownfield, the Bucket).

Thus, I think there should be an option on BucketRequest to configure the access policy on a per-namespace basis. And then I think BucketAccessClass is not needed at all.

Reply:
> Is this access policy about read and/or write access to certain prefixes within the bucket?

We have no direct support for [s3 prefixes](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-prefixes.html). The _BR.bucketPrefix_ field is a string that COSI will prepend to generated backend bucket names.

> there should be an option on BucketRequest to configure the access policy

Buckets can be shared within a namespace by creating multiple pods specifying the same BAR in their csi driver spec. Also, multiple BARs can reference the same BR (in the same ns) so allow for sharing with difference access policies.

Reply:
> Also, multiple BARs can reference the same BR (in the same ns) so allow for sharing with difference access policies.

Letting the users who create BARs configure the access policy is a security concern in multi-tenant. Instead, the user who creates the BR should be entitled to configure the access policy per-namespace. So I think:

1. `allowedNamespaces` should be added to `BucketRequest`.
1. `allowedNamespaces` is not flexible enough. It only allows to give another tenant full access or no access. Instead, it should be changed to allow configuring an access policy per namespace.

Reply:
> Not only "in the same ns", but in all ns contained in allowedNamespaces, right?

When access is needed outside of the BR's ns (this is for the greenfield -> brownfield use case) then the BAR needs to reference the B by name. This is because there is only 1 BR per B and it would be uncommon for a user to be able to reference a BR outside of their ns.

Reply:
What I said is regardless of whether the BucketAccessRequest references the Bucket or the BucketRequest.

IMO the creator of a BucketRequest should be able to configure the access policy per namespace. This configuration should be transfered from the BucketRequest to the Bucket. And it should be enforced for all accesses, regardless of whether the BucketAccessRequest references the Bucket or the BucketRequest.

Reply:
Perhaps you're thinking AccessPolicy is defined for a particular bucket prefix, rather than as a generic access policy. 

AccessPolicy must be defined as a generic access policy, i.e. `read/write:S3`, but not `read/write:S3:bucketName`. Then, the driver takes cares of associating the user with that bucket while granting privileges.

This way, the admin gets to prescribe the universe of all possible access policies (say for instance, admin wants to prevent object deletion for the entire org), and user gets to choose one type of access from it. The user still has control over what kind of access is obtained for their workload. 

I agree about `allowedNamespaces` not being flexible enough. This is a something I want to improve too. It is still under discussion. As of now, we are going with a primitive model to get things going before making improvements. Ideally, allowed namespaces should not be a property of the bucket, but rather it should be a parametrized value tied to the namespace, such that any namespace satisfying certain parameters can access the bucket. In addition, it should also encode the type of access allowed. You're welcome to join us at our future meetings to discuss this futher, but as of now - the current method is a good starting point.

Reply:
> This way, the admin gets to prescribe the universe of all possible access policies (say for instance, admin wants to prevent object deletion for the entire org), and user gets to choose one type of access from it.

A user can always create multiple BARs, one for each BAC, effectively obtaining access rights equal to the union of the prescribed access policies. Thus, a single BAC should always be sufficient.

But I see that an admin might want to have a different BAC for each BC. This raises the question: Why don't we simply move the fields from BAC to BC and remove BAC?